compiler/rustc_hir_typeck/src/demand.rs

```diff
@@ -1417,6 +1417,21 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         }
                     }
 
+                    if let Some(hir::Node::Expr(hir::Expr {
+                        kind: hir::ExprKind::Binary(_, left, right),
+                        ..
+                    })) = self.tcx.hir().find_parent(expr.hir_id) && mutability.is_mut() {
```

Comment:
I don't think this needs special-casing for mutable references.

Reply:
Do you mean `mutability.is_mut()`? It does to prevent this problem:

```rs
let a = String::new();
let b = String::new();
let _ = a + b // should suggestion a + &b, rather than *a + b
```

Reply:
Well what we really should be doing is checking whether the obligation `Lhs: BinOp<Rhs>` holds after dereffing the left-hand side. Not really sure if that's possible here, so I guess maybe give that a try or else I can approve..

Reply:
> checking whether the obligation Lhs: BinOp<Rhs> holds after dereffing the left-hand side

I haven't found a definitive solution yet. Do you have any thoughts or suggestions?